source/PlayerInfo.cpp

```diff
@@ -2354,7 +2364,34 @@ void PlayerInfo::StepMissions(UI *ui)
 		else if(mission.CanComplete(*this))
 			RemoveMission(Mission::COMPLETE, mission, ui);
 		else if(mission.Destination() == GetPlanet() && !freshlyLoaded)
+		{
 			mission.Do(Mission::VISIT, *this, ui);
+			if(mission.IsUnique())
+				continue;
+			
+			// On visit dialogs are handled separately as to avoid a player
+			// getting spammed by on visit dialogs if they are stacking jobs
+			// from the same destination.
+			auto action = mission.GetAction(Mission::VISIT);
+			if(!action.DialogText().empty() && visitText.empty())
+				visitText = Format::Replace(action.DialogText(), substitutions);
+			++missionVisits;
```

Comment:
I think we should avoid telling the player about any invisible missions that involve the current location.